### ISSUE TITLE
Error on trying to run detectron2 on Mac M1 with device set to "mps" (Metal Performance Shaders)

### ISSUE THREAD

**Issue by Bombardelli** [pytorch/pytorch#119968]
### 🐛 Describe the bug

I am trying to run detectron2 on Mac M1 with device set to "mps" (Metal Performance Shaders)

The problem is that every time I set up to mps the following error shows up:

Error: command buffer exited with error status.
	The Metal Performance Shaders operations encoded on it may not have completed.
	Error: 
	(null)
	Internal Error (0000000e:Internal Error)
	<AGXG13GFamilyCommandBuffer: 0x28bad31a0>
    label = <none> 
    device = <AGXG13GDevice: 0x14ed4ca00>
        name = Apple M1 
    commandQueue = <AGXG13GFamilyCommandQueue: 0x14fded000>
        label = <none> 
        device = <AGXG13GDevice: 0x14ed4ca00>
            name = Apple M1 
    retainedReferences = 1

Anyone else with the same problem? I believe there is currently an incompatibility of "mps" with detectron2. Because when you set to "cpu" it works normally, but it is cpu and therefore slower for training.

### Versions

Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 2606:50c0:8003::154, 2606:50c0:8001::154, 2606:50c0:8002::154, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|2606:50c0:8003::154|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 22068 (22K) [text/plain]
Saving to: ‘collect_env.py’

collect_env.py      100%[===================>]  21.55K  --.-KB/s    in 0.001s  

2024-02-14 22:32:30 (20.7 MB/s) - ‘collect_env.py’ saved [22068/22068]

zsh: command not found: #
Collecting environment information...
PyTorch version: 2.2.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.2.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.1.0.2.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.13 | packaged by conda-forge | (main, Oct 26 2023, 18:09:17) [Clang 16.0.6 ] (64-bit runtime)
Python platform: macOS-14.2.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] torch==2.2.0
[pip3] torchaudio==2.2.0
[pip3] torchvision==0.17.0
[conda] numpy                     1.26.4          py310hd45542a_0    conda-forge
[conda] pytorch                   2.2.0                  py3.10_0    pytorch
[conda] torch                     2.2.0                    pypi_0    pypi
[conda] torchaudio                2.2.0                    pypi_0    pypi
[conda] torchvision               0.17.0                   pypi_0    pypi

**Comment by jamorphy** [comment]
Running into the same error when running `fasterrcnn_resnet50_fpn`, also on M1:
```
Error: command buffer exited with error status.
	The Metal Performance Shaders operations encoded on it may not have completed.
	Error: 
	(null)
	Internal Error (0000000e:Internal Error)
	<AGXG13XFamilyCommandBuffer: 0x2ddd4bc60>
    label = <none> 
    device = <AGXG13XDevice: 0x12566b200>
        name = Apple M1 Max 
    commandQueue = <AGXG13XFamilyCommandQueue: 0x130cc4400>
        label = <none> 
        device = <AGXG13XDevice: 0x12566b200>
            name = Apple M1 Max 
    retainedReferences = 1
```

```
numpy                                        1.24.4
torch                                        2.2.0
torchaudio                                   2.2.0
torchvision                                  0.17.0
```

**Comment by dotpyo** [comment]
Hello,

> TL;DR: It is OOM 
https://github.com/google/jax/issues/19110#issuecomment-1868385445

I had same issue on my project. but it becomes to work when I cut down `batch size`
literally, I searched the line `Error: command buffer exited with error status.` around every google, github, pytorch forum
they commonly say that reducing batch size was work in some how. VRAM issue is the main reason I suspect. no one knew the actual reason but their experience have suffcient grounds.

hope this would help

**Comment by claganier** [comment]
Same issue with `fasterrcnn_resnet50_fpn` on M1 Pro with a batch size of 2 and 800px images.

```
numpy                                        1.26.0
torch                                        2.1.0
torchaudio                                   2.1.0
torchvision                                  0.16.0
```

Any planned fixes or ways around this ?

**Comment by kofiarkoh** [comment]
same issue with  `fasterrcnn_resnet50_fpn`
<img width="759" alt="Screenshot 2024-05-19 at 10 25 03 PM" src="https://github.com/pytorch/pytorch/assets/47466360/344ca44f-7d7e-490a-95b1-927a4a069384">

**Comment by Noah-Yarmush** [comment]
I have a similar error for **device = "mps"** and training **torchvision.models.detection.fasterrcnn_resnet50_fpn** network. However, when I hardly code **device = "cpu"** it performs fast (few seconds) and there's no error message but I cannot say at this moment if the training goes successfully and is not stopped internally without errors.

**Comment by Kacper0199** [comment]
I am encountering the same issue while running `fasterrcnn_resnet50_fpn` with `FastRCNNPredictor` on my MacBook M1 Pro using the "mps" device (Metal Performance Shaders). The error appears multiple times during training with the following message:

```
Error: command buffer exited with error status.
The Metal Performance Shaders operations encoded on it may not have completed.
Error:
(null)
Internal Error (0000000e:Internal Error)
<AGXG13XFamilyCommandBuffer: 0x4baef3030>
label = <none>
device = <AGXG13XDevice: 0x30c52b600>
name = Apple M1 Pro
commandQueue = <AGXG13XFamilyCommandQueue: 0x12f870800>
label = <none>
device = <AGXG13XDevice: 0x30c52b600>
name = Apple M1 Pro
retainedReferences = 1
```

I am using the following setup to train a Faster R-CNN model with 2 classes on 800px images:

```python
def get_model(num_classes):
    model = torchvision.models.detection.fasterrcnn_resnet50_fpn(weights='DEFAULT')
    in_features = model.roi_heads.box_predictor.cls_score.in_features
    model.roi_heads.box_predictor = FastRCNNPredictor(in_features, num_classes)
    return model

num_classes = 2
model = get_model(num_classes)
model.to(device)
```

The error occurs at random points during training, typically after a few iterations, and sometimes freezes the system.

Here is my environment:

```
PyTorch: 2.4.1
Torchvision: 0.19.1
NumPy: 1.26.4
Device: MacBook M1 Pro
Batch size: 4 and more
```

Like others, I found that training on cpu works, but it’s slower. Reducing the batch size also delays the issue but doesn’t fully resolve it, slowing training process. I believe there is still some incompatibility with mps and fasterrcnn_resnet50_fpn ...

Has anyone found a reliable workaround or fix for this issue?

**Comment by t-mdo** [comment]
I have the same issue using `maskrcnn_resnet50_fpn` during inference, and it's done on a batch of 1.

**Comment by andrenatal** [comment]
I'm having same issue on `maskrcnn_resnet50_fpn_v2` on Apple M2 Max. Sad

```
Error: command buffer exited with error status.
        The Metal Performance Shaders operations encoded on it may not have completed.
        Error: 
        (null)
        Internal Error (0000000e:Internal Error)
        <AGXG14XFamilyCommandBuffer: 0x39d9f19d0>
    label = <none> 
    device = <AGXG14CDevice: 0x1192c4e00>
        name = Apple M2 Max 
    commandQueue = <AGXG14XFamilyCommandQueue: 0x1192c6400>
        label = <none> 
        device = <AGXG14CDevice: 0x1192c4e00>
            name = Apple M2 Max 
    retainedReferences = 1
Error: command buffer exited with error status.
        The Metal Performance Shaders operations encoded on it may not have completed.
        Error: 
        (null)
        Internal Error (0000000e:Internal Error)
        <AGXG14XFamilyCommandBuffer: 0x39d9d4ff0>
    label = <none> 
    device = <AGXG14CDevice: 0x1192c4e00>
        name = Apple M2 Max 
    commandQueue = <AGXG14XFamilyCommandQueue: 0x1192c6400>
        label = <none> 
        device = <AGXG14CDevice: 0x1192c4e00>
            name = Apple M2
```
